### DISCUSSION TITLE
How to set (connection_)collation for mysqldb after commit f5e57f7c31?

### DISCUSSION THREAD

**Discussion by luchr** [sqlalchemy/sqlalchemy#7858] [Usage Questions] [answered]
Hello,

I have a problem since commit f5e57f7c31:  SqlAlchemy now always emits `SET NAMES` for new connections in the mysqldb dialect. This resets the collation to the **compile(!)-time** default of the server, see [mariadb documentation: set names](https://mariadb.com/kb/en/set-names/). In my case the character encoding ist `utf8mb4` and I wish to have the collation `utf8mb4_unicode_520_ci` (because this is the server collation and the database collation, etc.). So I tried to create the engine with
```
{
  "url": "mysql+mysqldb://user:****@host/db?charset=utf8mb4",
  "connect_args": {
    "init_command": "SET SESSION collation_connection='utf8mb4_unicode_520_ci'"
  }
}
```
But now with the emitted `SET NAMES utf8mb4` the `collation_connection` is set back to `utf8mb4_general_ci`. Hence the `init_command` is useless, because the effect only last until the `SET NAMES` thing. So I have to manually do
```
SET NAMES utf8mb4 COLLATE utf8mb4_unicode_520_ci;
```
after the connection to get back the wanted collation.

Is there a way to specify the wanted collation somewhere?

**Comment by zzzeek** [comment, marked as answer]
it seems like our SET NAMES should also look up the collation and set that also, so that nothing is reset.   if you can suggest a modification to f5e57f7c311288d892894edcc44d901b5bfbb3d1 which achieves this we can commit it.
 
however for now you can use a pool [connect](https://docs.sqlalchemy.org/en/14/core/events.html?highlight=poolevents%20connect#sqlalchemy.events.PoolEvents.connect) event to set this parameter back.  it *should* be invoked *after* the call that the dialect is doing.    let me know if it's not doing that.

**Reply by luchr** [reply]
Thank you for your fast reply. Yes, the `connect` event for the engine is working. I use it to (re-)execute the `init_command` (a second time). That's a nice workaround  :-)

I dont' get your idea of "look up the collation". My wish to tackle this is an optional connect parameter `collation`:
```
"mysql+mysqldb://user:****@host/db?charset=utf8mb4&collation=utf8mb4_unicode_520_ci"
```
If it is specified then the long `SET NAMES <charset> COLLATE <collation>` is used, otherwise the old behaviour `SET NAMES <charset>`.

Do you mean with "look up the collation" to check for the collation before the `SET NAMES` and try to use `SET NAMES <charset> COLLATE <last found collation>`? I guess, this won't work for every case.

**Reply by zzzeek** [reply]
the code that was added, which I added because this is some driver thing on the MySQL side that seems to be needed, retrieves the "charset" from the connection, then feeds it into SET NAMES:

```py
   charset_name = conn.character_set_name()

   if charset_name is not None:
            cursor = conn.cursor()
            cursor.execute("SET NAMES %s" % charset_name)
            cursor.close()
```

totally seems like driver stuff to me, but anyway.  So we are saying, "SET NAMES" is blowing away your collation setting.  So my proposal is to change the above to this:

```py
   charset_name = conn.character_set_name()
   collation_name = conn.<??????>

   if charset_name is not None:
            cursor = conn.cursor()
            cursor.execute("SET NAMES %s COLLATE %s" % (charset_name, collation_name)
            cursor.close()
```

where the `?????` part is where we get what should be set there.

**Reply by luchr** [reply]
I see only two possible sources/candidates for getting the `??????`: 

1. The connection/SQL: But how to know if the collation *before* the `SET NAMES` matches to the charset that is used in `SET NAMES`? There is a reason for this `SET NAMES` (the reason for the mentioned commit). So just reading the collation (e.g. `SELECT @collation_connection`) may horrible fail: may be `NULL` or may be completely wrong like in the case where `SET NAMES` was invented (the reason for this commit), because the charset/collation before was not suitable.
2. Some (Config-)Variables in Python/SqlAlchemy: That was (part of) my question. I don't know of such an (important) information attached to a connection. [As I stated above charset and collation belong together. In SQL for types one can specify charset and collation, together. I don't know why SqlAlchemy is treating them differently. For charset there is a `charset` parameter, but for collation one has to use some `connect_args`, `init_command` things.]

So all candidates, I am aware of, won't work for all cases (in my opinion).

**Reply by zzzeek** [reply]
i wish the driver would just fix this.   so for now, do the workaround.
